JavadocMethod: new property 'allowInlineReturn' to support for Javadoc return tag as inline

I have downloaded the latest cli from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

**How it works Now:**

```bash
$ javac Foo.java
$ javadoc Foo.java 
Loading source file Foo.java...
Constructing Javadoc information...
Building index for all the packages and classes...
Standard Doclet version 16+36
Building tree for all the packages and classes...
Generating ./Foo.html...
Generating ./package-summary.html...
Generating ./package-tree.html...
Generating ./overview-tree.html...
Building index for all classes...
Generating ./allclasses-index.html...
Generating ./allpackages-index.html...
Generating ./index-all.html...
Generating ./index.html...
Generating ./help-doc.html...
```

config.xml
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
   <property name="fileExtensions" value="java"/>
   <module name="TreeWalker">
       <module name="JavadocMethod">
            <property name="accessModifiers" value="public"/>
       </module>
   </module>
</module>
```

Foo.java
```java
/**
 * Test
 */
public class Foo {

    /**
     * {@return the foo}
     */
    public int getFoo() { return 0; } // violation

    /**
     * Returns the bar
     * @return the bar
     */
    public int getBar() { return 0; }

}
```

```bash
$ RUN_LOCALE="-Duser.language=en -Duser.country=US"
$ java $RUN_LOCALE -jar ~/Downloads/checkstyle-9.0-all.jar -c config.xml Foo.java
Starting audit...
[ERROR] /tmp/Foo.java:9: @return tag should be present and have description. [JavadocMethod]
Audit done.
Checkstyle ends with 1 errors.
```

**Is your feature request related to a problem? Please describe.**
Add support for Javadoc inline return tag

**Describe the solution you'd like**
Both methods to be treated the same (getFoo and getBar), as it is by the Javadoc

**Additional context**
The support for this syntax was added in JDK 16
https://bugs.openjdk.java.net/browse/JDK-8075778

```
rivanov@p5510:/var/tmp$ javadoc Foo.java
Loading source file Foo.java...
Constructing Javadoc information...
Standard Doclet version 11
Building tree for all the packages and classes...
Generating ./Foo.html...
Foo.java:7: error: no tag name after @
     * {@return the foo}
       ^
Foo.java:9: warning: no @return
    public int getFoo() { return 0; } // violation
               ^


/var/tmp$ export JAVA_HOME=/opt/jvm/jdk-17.0.1/
/var/tmp$ export PATH=$JAVA_HOME/bin:$PATH
rivanov@p5510:/var/tmp$ javadoc Foo.java
Loading source file Foo.java...
Constructing Javadoc information...
Building index for all the packages and classes...
Standard Doclet version 17.0.1+12-39
Building tree for all the packages and classes...
Generating ./Foo.html...

```
![image](https://user-images.githubusercontent.com/812984/147023606-4df25f74-666d-4c53-933d-fcba6a9d7ff5.png)

@romani Here is the requested information.
http://rveach.no-ip.org/checkstyle/regression/9745/

````
$ /usr/lib/jvm/java-16-openjdk/bin/javadoc -d 16 ./Foo.java
Loading source file ./Foo.java...
Constructing Javadoc information...
Creating destination directory: "16/"
Building index for all the packages and classes...
Standard Doclet version 16.0.2+7-67
Building tree for all the packages and classes...
Generating 16/Foo.html...
Generating 16/package-summary.html...
Generating 16/package-tree.html...
Generating 16/overview-tree.html...
Building index for all classes...
Generating 16/allclasses-index.html...
Generating 16/allpackages-index.html...
Generating 16/index-all.html...
Generating 16/index.html...
Generating 16/help-doc.html...

$ javadoc -d 11 ./Foo.java
Loading source file ./Foo.java...
Constructing Javadoc information...
Creating destination directory: "11/"
Standard Doclet version 11.0.11
Building tree for all the packages and classes...
Generating 11/Foo.html...
./Foo.java:7: warning - invalid usage of tag {@return the foo}
./Foo.java:7: warning - invalid usage of tag {@return the foo}
Generating 11/package-summary.html...
Generating 11/package-tree.html...
Generating 11/constant-values.html...
Building index for all the packages and classes...
Generating 11/overview-tree.html...
Generating 11/index-all.html...
./Foo.java:7: warning - invalid usage of tag {@return the foo}
Building index for all classes...
Generating 11/allclasses-index.html...
Generating 11/allpackages-index.html...
Generating 11/deprecated-list.html...
Building index for all classes...
Generating 11/allclasses.html...
Generating 11/allclasses.html...
Generating 11/index.html...
Generating 11/help-doc.html...
3 warnings
````

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java
@@ -36,7 +36,9 @@ import com.puppycrawl.tools.checkstyle.utils.JavadocUtil;
  * <a href="https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html#firstsentence">
  * Javadoc summary sentence</a> does not contain phrases that are not recommended to use.
  * Summaries that contain only the {@code {@inheritDoc}} tag are skipped.
- * Check also violate Javadoc that does not contain first sentence.
+ * Summaries that contain a non-empty {@code {@return}} are allowed.
+ * Check also violate Javadoc that does not contain first sentence, though with {@code {@return}} a
+ * period is not required as the Javadoc tool adds it.
  * </p>
  * <ul>
  * <li>
@@ -255,6 +257,9 @@ public class SummaryJavadocCheck extends AbstractJavadocCheck {
     /** Summary tag text. */
     private static final String SUMMARY_TEXT = "@summary";
 
+    /** Return tag text. */
+    private static final String RETURN_TEXT = "@return";
+
     /** Set of allowed Tokens tags in summary java doc. */
     private static final Set<Integer> ALLOWED_TYPES = Set.of(
                     JavadocTokenTypes.WS,
@@ -303,46 +308,52 @@ public class SummaryJavadocCheck extends AbstractJavadocCheck {
 
     @Override
     public void visitJavadocToken(DetailNode ast) {
-        final Optional<DetailNode> inlineSummaryTag = getInlineSummaryTag(ast);
-        if (inlineSummaryTag.isPresent()) {
-            validateSummaryTag(inlineSummaryTag.get());
+        final Optional<DetailNode> inlineTag = getInlineTagNode(ast);
+        if (inlineTag.isPresent() && isSummaryTag(inlineTag.get())) {
+            validateSummaryTag(inlineTag.get());
+        }
+        else if (inlineTag.isPresent() && isInlineReturnTag(inlineTag.get())) {
+            validateInlineReturnTag(inlineTag.get());
         }
         else if (!startsWithInheritDoc(ast)) {
-            final String summaryDoc = getSummarySentence(ast);
-            if (summaryDoc.isEmpty()) {
-                log(ast.getLineNumber(), MSG_SUMMARY_JAVADOC_MISSING);
+            validateUntaggedSummary(ast);
+        }
+    }
+
+    /**
+     * Checks the javadoc text for {@code period} at end and forbidden fragments.
+     *
+     * @param ast the javadoc text node
+     */
+    private void validateUntaggedSummary(DetailNode ast) {
+        final String summaryDoc = getSummarySentence(ast);
+        if (summaryDoc.isEmpty()) {
+            log(ast.getLineNumber(), MSG_SUMMARY_JAVADOC_MISSING);
+        }
+        else if (!period.isEmpty()) {
+            final String firstSentence = getFirstSentence(ast);
+            final int endOfSentence = firstSentence.lastIndexOf(period);
+            if (!summaryDoc.contains(period)) {
+                log(ast.getLineNumber(), MSG_SUMMARY_FIRST_SENTENCE);
             }
-            else if (!period.isEmpty()) {
-                final String firstSentence = getFirstSentence(ast);
-                final int endOfSentence = firstSentence.lastIndexOf(period);
-                if (!summaryDoc.contains(period)) {
-                    log(ast.getLineNumber(), MSG_SUMMARY_FIRST_SENTENCE);
-                }
-                if (endOfSentence != -1
-                        && containsForbiddenFragment(firstSentence.substring(0, endOfSentence))) {
-                    log(ast.getLineNumber(), MSG_SUMMARY_JAVADOC);
-                }
+            if (endOfSentence != -1
+                    && containsForbiddenFragment(firstSentence.substring(0, endOfSentence))) {
+                log(ast.getLineNumber(), MSG_SUMMARY_JAVADOC);
             }
         }
     }
 
     /**
-     * Gets the inline summary tag.
+     * Gets the node for the inline tag if present.
      *
      * @param javadoc javadoc root node.
-     * @return an optional of inline summary tag node, empty optional if inline tag is not
-     *         a summary tag.
+     * @return the node for the inline tag if present.
      */
-    private static Optional<DetailNode> getInlineSummaryTag(DetailNode javadoc) {
-        Optional<DetailNode> node = Arrays.stream(javadoc.getChildren())
-                .filter(SummaryJavadocCheck::isInlineTagPresent)
-                .findFirst()
-                .map(SummaryJavadocCheck::getInlineTagNodeWithinHtmlElement);
-
-        if (node.isPresent() && !isSummaryTag(node.get())) {
-            node = Optional.empty();
-        }
-        return node;
+    private static Optional<DetailNode> getInlineTagNode(DetailNode javadoc) {
+        return Arrays.stream(javadoc.getChildren())
+            .filter(SummaryJavadocCheck::isInlineTagPresent)
+            .findFirst()
+            .map(SummaryJavadocCheck::getInlineTagNodeWithinHtmlElement);
     }
 
     /**
@@ -392,13 +403,35 @@ public class SummaryJavadocCheck extends AbstractJavadocCheck {
      * @return {@code true} if inline tag is summary tag.
      */
     private static boolean isSummaryTag(DetailNode javadocInlineTag) {
+        return isInlineTagWithName(javadocInlineTag, SUMMARY_TEXT);
+    }
+
+    /**
+     * Checks if the first tag inside ast is {@code {@return}} tag.
+     *
+     * @param javadocInlineTag node of type {@link JavadocTokenTypes#JAVADOC_INLINE_TAG}
+     * @return {@code true} if first tag is return tag.
+     */
+    private static boolean isInlineReturnTag(DetailNode javadocInlineTag) {
+        return isInlineTagWithName(javadocInlineTag, RETURN_TEXT);
+    }
+
+    /**
+     * Checks if the first tag inside ast is a tag with the given name.
+     *
+     * @param javadocInlineTag node of type {@link JavadocTokenTypes#JAVADOC_INLINE_TAG}
+     * @param name name of inline tag.
+     *
+     * @return {@code true} if first tag is a tag with the given name.
+     */
+    private static boolean isInlineTagWithName(DetailNode javadocInlineTag, String name) {
         final DetailNode[] child = javadocInlineTag.getChildren();
 
         // Checking size of ast is not required, since ast contains
         // children of Inline Tag, as at least 2 children will be present which are
         // RCURLY and LCURLY.
         return child[1].getType() == JavadocTokenTypes.CUSTOM_NAME
-                && SUMMARY_TEXT.equals(child[1].getText());
+            && name.equals(child[1].getText());
     }
 
     /**
@@ -419,6 +452,22 @@ public class SummaryJavadocCheck extends AbstractJavadocCheck {
             else if (containsForbiddenFragment(inlineSummary)) {
                 log(inlineSummaryTag.getLineNumber(), MSG_SUMMARY_JAVADOC);
             }
+        }
+    }
+
+    /**
+     * Checks the inline return for forbidden fragments.
+     *
+     * @param inlineReturnTag node of type {@link JavadocTokenTypes#JAVADOC_INLINE_TAG}
+     */
+    private void validateInlineReturnTag(DetailNode inlineReturnTag) {
+        final String inlineReturn = getContentOfInlineCustomTag(inlineReturnTag);
+        final String returnVisible = getVisibleContent(inlineReturn);
+        if (returnVisible.isEmpty()) {
+            log(inlineReturnTag.getLineNumber(), MSG_SUMMARY_JAVADOC_MISSING);
+        }
+        else if (containsForbiddenFragment(inlineReturn)) {
+            log(inlineReturnTag.getLineNumber(), MSG_SUMMARY_JAVADOC);
         }
     }
 

--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/javadoc/SummaryJavadocCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/javadoc/SummaryJavadocCheck.xml
@@ -9,7 +9,9 @@
  &lt;a href="https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html#firstsentence"&gt;
  Javadoc summary sentence&lt;/a&gt; does not contain phrases that are not recommended to use.
  Summaries that contain only the {@code {@inheritDoc}} tag are skipped.
- Check also violate Javadoc that does not contain first sentence.
+ Summaries that contain a non-empty {@code {@return}} are allowed.
+ Check also violate Javadoc that does not contain first sentence, though with {@code {@return}} a
+ period is not required as the Javadoc tool adds it.
  &lt;/p&gt;</description>
          <properties>
             <property default-value="false"

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheckTest.java
@@ -172,6 +172,28 @@ public class SummaryJavadocCheckTest extends AbstractModuleTestSupport {
     }
 
     @Test
+    public void testInlineReturn() throws Exception {
+        final String[] expected = {
+            "74: " + getCheckMessage(MSG_SUMMARY_JAVADOC_MISSING),
+        };
+
+        verifyWithInlineConfigParser(
+                getPath("InputSummaryJavadocInlineReturn.java"), expected);
+    }
+
+    @Test
+    public void testInlineReturnForbidden() throws Exception {
+        final String[] expected = {
+            "14: " + getCheckMessage(MSG_SUMMARY_JAVADOC),
+            "21: " + getCheckMessage(MSG_SUMMARY_JAVADOC),
+            "28: " + getCheckMessage(MSG_SUMMARY_JAVADOC),
+        };
+
+        verifyWithInlineConfigParser(
+                getPath("InputSummaryJavadocInlineReturnForbidden.java"), expected);
+    }
+
+    @Test
     public void testPeriodAtEnd() throws Exception {
         final String[] expected = {
             "19: " + getCheckMessage(MSG_SUMMARY_JAVADOC_MISSING),

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/summaryjavadoc/InputSummaryJavadocInlineReturn.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/summaryjavadoc/InputSummaryJavadocInlineReturn.java
@@ -1,0 +1,89 @@
+/*
+SummaryJavadoc
+violateExecutionOnNonTightHtml = (default)false
+forbiddenSummaryFragments = (default)^$
+period = (default).
+
+*/
+
+package com.puppycrawl.tools.checkstyle.checks.javadoc.summaryjavadoc;
+
+class InputSummaryJavadocInlineReturn {
+    /**
+     * {@return a zero, note that no period is ok as it is set by javadoc tool}
+     */
+    private int returnFoo1() // ok
+    {
+        return 0;
+    }
+
+    /**
+     * {@return a {@code 0}}
+     */
+    private int returnFoo2() // ok
+    {
+        return 0;
+    }
+
+    /**
+     * Text before the return tag. {@return a {@code 0}}
+     */
+    private int returnFoo3() // ok, javadoc tool produces a warning
+    {
+        return 0;
+    }
+
+    /**
+     * {@return a {@code 0}} Text after the return tag.
+     */
+    private int returnFoo4() // ok
+    {
+        return 0;
+    }
+
+    /**
+     * {@return a {@code 0}
+     * there's more text on another line
+     * and even more lines}
+     */
+    private int returnMultiline() // ok
+    {
+        return 0;
+    }
+
+    /**
+     * {@return a {@code 0}<br>
+     * there's more text on another line<br>
+     * and even more lines}
+     */
+    private int returnMultiline2() // ok
+    {
+        return 0;
+    }
+
+    /**
+     * {@return nothing, this is a void method}
+     */
+    private void voidMethod() // ok, javadoc tool produces an error
+    {
+    }
+
+    // the return tag is empty, which is not allowed
+    /**
+     // violation below
+     * {@return}
+     */
+    int returnNothing() {
+        return 0;
+    }
+
+    /**
+     * {@return nothing, this is a field}
+     */
+    private static final byte NOT_A_METHOD = 0; // ok, javadoc tool produces an error
+
+    /**
+     * {@return nothing, this is a class}
+     */
+    private class NotAMethod {} // ok, javadoc tool produces an error
+}

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/summaryjavadoc/InputSummaryJavadocInlineReturnForbidden.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/summaryjavadoc/InputSummaryJavadocInlineReturnForbidden.java
@@ -1,0 +1,33 @@
+/*
+SummaryJavadoc
+violateExecutionOnNonTightHtml = (default)false
+forbiddenSummaryFragments = the .*|This method returns
+period = (default).
+
+*/
+
+package com.puppycrawl.tools.checkstyle.checks.javadoc.summaryjavadoc;
+
+public class InputSummaryJavadocInlineReturnForbidden {
+    /**
+     // violation below
+     * {@return the nothing}
+     */
+    int returnNothing() {
+        return 0;
+    }
+    /**
+     // violation below
+     * {@return This method returns something}
+     */
+    int returnSomething() {
+        return 0;
+    }
+    /**
+     // violation below
+     * {@return This method returns something.}
+     */
+    int returnSomethingElse() {
+        return 0;
+    }
+}

--- a/src/xdocs/config_javadoc.xml
+++ b/src/xdocs/config_javadoc.xml
@@ -3619,8 +3619,10 @@ public int foobar() {
           Checks that
           <a href="https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html#firstsentence">
           Javadoc summary sentence</a> does not contain phrases that are not recommended to use.
-          Summaries that contain only the <code>{@inheritDoc}</code> tag are skipped. Check also
-          violate Javadoc that does not contain first sentence.
+          Summaries that contain only the <code>{@inheritDoc}</code> tag are skipped. Summaries
+          that contain a non-empty {@code {@return}} are allowed. Check also violate Javadoc that
+          does not contain first sentence, though with {@code {@return}} a period is not required
+          as the Javadoc tool adds it.
         </p>
       </subsection>
 

